core/block/detailservice/relations.go

```diff
@@ -0,0 +1,146 @@
+package detailservice
+
+import (
+	"context"
+	"fmt"
+	"slices"
+	"strings"
+	"time"
+
+	"github.com/gogo/protobuf/types"
+	"go.uber.org/zap"
+	"golang.org/x/exp/maps"
+
+	"github.com/anyproto/anytype-heart/core/block/cache"
+	"github.com/anyproto/anytype-heart/core/block/editor/smartblock"
+	"github.com/anyproto/anytype-heart/core/domain"
+	"github.com/anyproto/anytype-heart/pkg/lib/bundle"
+	coresb "github.com/anyproto/anytype-heart/pkg/lib/core/smartblock"
+	"github.com/anyproto/anytype-heart/pkg/lib/database"
+	"github.com/anyproto/anytype-heart/pkg/lib/localstore/addr"
+	"github.com/anyproto/anytype-heart/pkg/lib/pb/model"
+	"github.com/anyproto/anytype-heart/space/spacecore/typeprovider"
+	"github.com/anyproto/anytype-heart/util/pbtypes"
+	"github.com/anyproto/anytype-heart/util/slice"
+)
+
+var ErrBundledTypeIsReadonly = fmt.Errorf("can't modify bundled object type")
+
+func (s *service) ObjectTypeAddRelations(ctx context.Context, objectTypeId string, relationKeys []domain.RelationKey) error {
+	if strings.HasPrefix(objectTypeId, bundle.TypePrefix) {
+		return ErrBundledTypeIsReadonly
+	}
+	return cache.Do(s, objectTypeId, func(b smartblock.SmartBlock) error {
+		st := b.NewState()
+		list := pbtypes.GetStringList(st.Details(), bundle.RelationKeyRecommendedRelations.String())
+		for _, relKey := range relationKeys {
+			relId, err := b.Space().GetRelationIdByKey(ctx, relKey)
+			if err != nil {
+				return err
+			}
+			if !slices.Contains(list, relId) {
+				list = append(list, relId)
+			}
+		}
+		st.SetDetailAndBundledRelation(bundle.RelationKeyRecommendedRelations, pbtypes.StringList(list))
+		return b.Apply(st)
+	})
+}
+
+func (s *service) ObjectTypeRemoveRelations(ctx context.Context, objectTypeId string, relationKeys []domain.RelationKey) error {
+	if strings.HasPrefix(objectTypeId, bundle.TypePrefix) {
+		return ErrBundledTypeIsReadonly
+	}
+	return cache.Do(s, objectTypeId, func(b smartblock.SmartBlock) error {
+		st := b.NewState()
+		list := pbtypes.GetStringList(st.Details(), bundle.RelationKeyRecommendedRelations.String())
+		for _, relKey := range relationKeys {
+			relId, err := b.Space().GetRelationIdByKey(ctx, relKey)
+			if err != nil {
+				return fmt.Errorf("get relation id by key %s: %w", relKey, err)
+			}
+			list = slice.RemoveMut(list, relId)
+		}
+		st.SetDetailAndBundledRelation(bundle.RelationKeyRecommendedRelations, pbtypes.StringList(list))
+		return b.Apply(st)
+	})
+}
+
+func (s *service) ListRelationsWithValue(spaceId string, value *types.Value) (keys []string, counters []int64, err error) {
+	countersByKeys := make(map[string]int64)
+	detailHandlesValue := generateFilter(value)
+
+	err = s.store.QueryIterate(database.Query{Filters: []*model.BlockContentDataviewFilter{{
+		RelationKey: bundle.RelationKeySpaceId.String(),
+		Condition:   model.BlockContentDataviewFilter_Equal,
+		Value:       pbtypes.String(spaceId),
+	}}}, func(details *types.Struct) {
+		for key, v := range details.Fields {
```

Comment:
Can we rename v to value or something? Because it looks like not consistent naming 🧐

Reply:
Yeah, fixed